modin/engines/base/frame/data.py

```diff
@@ -1632,7 +1632,23 @@ def _copartition(self, axis, other, how, sort, force_repartition=False):
         """
         if isinstance(other, type(self)):
             other = [other]
-        if all(o.axes[axis].equals(self.axes[axis]) for o in other):
+
+        is_partitons_aligned = True
+        if len(self._partitions) != len(other[0]._partitions) and len(
```

Comment:
I think the condition is confusing and incorrect a little bit. Imagine if `axis=1` and `len(self.axes[axis]) == len(other[0].axes[axis])` is True, but number of partitions is different in `self` and `other` on `axis=1`.

Reply:
Do you mean in `len(self._partitions) != len(other[0]._partitions)` I check only by `axis=0`?

Reply:
Yes, I mean this.

Reply:
I didn't find any situations when we should make re-partitioning in case when we have not aligned col-partitions. With parameter `axis=1` we will use `_copartition` only when `type(self)` is dataframe and `type(other)` is series. I this case repartitioning isn't needed.

Reply:
A situation when repartitioning is needed for axis=1:
```
import modin.pandas as pd
import pandas
data = [0, 1, 2, 3, 4, 5]
mdf1 = pd.DataFrame({"a": data, "b": data})
mdf2 = pd.concat((pd.Series(data[:3]), pd.Series(data[:3], index=[3,4,5])))
mdf1 + mdf2
ValueError: cannot reindex from a duplicate axis
pdf1 = pandas.DataFrame({"a": data, "b": data})
pdf2 = pandas.concat((pandas.Series(data[:3]), pandas.Series(data[:3], index=[3,4,5])))
pdf1 + pdf2
    a   b   0   1   2   3   4   5
0 NaN NaN NaN NaN NaN NaN NaN NaN
1 NaN NaN NaN NaN NaN NaN NaN NaN
2 NaN NaN NaN NaN NaN NaN NaN NaN
3 NaN NaN NaN NaN NaN NaN NaN NaN
4 NaN NaN NaN NaN NaN NaN NaN NaN
5 NaN NaN NaN NaN NaN NaN NaN NaN

Reply:
This code successfully works on this branch. Do you run using actual revision?

Reply:
Sorry, I used not last revision. Anyway, the results are different:
```python
modin result
    a   b
0 NaN NaN
1 NaN NaN
2 NaN NaN
3 NaN NaN
4 NaN NaN
5 NaN NaN
pandas_result
    a   b   0   1   2   3   4   5
0 NaN NaN NaN NaN NaN NaN NaN NaN
1 NaN NaN NaN NaN NaN NaN NaN NaN
2 NaN NaN NaN NaN NaN NaN NaN NaN
3 NaN NaN NaN NaN NaN NaN NaN NaN
4 NaN NaN NaN NaN NaN NaN NaN NaN
5 NaN NaN NaN NaN NaN NaN NaN NaN
```

Reply:
Hm, it's strange. My runs give me correct results.
```
modin 
     a   b   0   1   2   3   4   5
0 NaN NaN NaN NaN NaN NaN NaN NaN
1 NaN NaN NaN NaN NaN NaN NaN NaN
2 NaN NaN NaN NaN NaN NaN NaN NaN
3 NaN NaN NaN NaN NaN NaN NaN NaN
4 NaN NaN NaN NaN NaN NaN NaN NaN
5 NaN NaN NaN NaN NaN NaN NaN NaN
pandas 
     a   b   0   1   2   3   4   5
0 NaN NaN NaN NaN NaN NaN NaN NaN
1 NaN NaN NaN NaN NaN NaN NaN NaN
2 NaN NaN NaN NaN NaN NaN NaN NaN
3 NaN NaN NaN NaN NaN NaN NaN NaN
4 NaN NaN NaN NaN NaN NaN NaN NaN
5 NaN NaN NaN NaN NaN NaN NaN NaN
```

Reply:
Looks like for Ray results are correct, but for Dask is not correct.

Reply:
Discussed 1:1, the case is working.